File: src/main/java/com/dsa/problems/scaler/linked_list/kth_node.java
package com.dsa.problems.scaler.linked_list;

import java.util.ArrayList;
import java.util.Arrays;

public class kth_node {
  /**
   * Q. kth Node in a List
   *
   * You are given a singly linked list and an integer k. Your task is to access the node at the k-th index (0-based indexing) in the list and return its value. If the index is out of bounds, return -1.
   *
   * Problem Constraints
   * k<=10^6
   */
  private static int kthNode(Node A, int B) {
    if(B == 0) {
      return A.data;
    }
    Node temp = A;
    for(int i = 1; i < B && temp.next != null; i++) {
      temp = temp.next;
    }
    if(temp.next != null) 
      return temp.next.data;
    else return -1;
  }

  public static void main(String[] args) {
    Node temp = create_ll.createLL(new ArrayList<>(Arrays.asList(1,2,3,4,5,6)));
    System.out.println(kthNode(temp, 0));
    System.out.println(kthNode(temp, 3));
    System.out.println(kthNode(temp, 30));
    System.out.println(kthNode(temp, 5));
  }
}

File: problems/scaler/linked_list/kth_node.js
/**
 * Q. kth Node in a List
 *
 * You are given a singly linked list and an integer k. Your task is to access the node at the k-th index (0-based indexing) in the list and return its value. If the index is out of bounds, return -1.
 *
 * Problem Constraints
 * k<=10^6
 */

// Definition for singly-linked list.
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

function createLL(A) {
  let head = null;
  let temp = null;
  for (let a of A) {
    const newNode = new Node(a);
    if (head === null) {
      head = newNode;
      temp = newNode;
    } else {
      temp.next = newNode;
      temp = temp.next;
    }
  }
  return head;
}

function soln(A, B) {
  let temp = A;
  while (B-- && temp.next) {
    temp = temp.next;
  }
  if (B >= 0) {
    return -1;
  }
  return temp.data;
}

const head = createLL([
  90, 14, 83, 66, 45, 26, 29, 14, 52, 24, 29, 22, 97, 95, 98, 65, 90, 98, 51,
  80, 62, 10, 66, 86, 57,
]);
soln(head, 18);
